src/agent/restart_handler/src/restart_handler_unix.cpp

```diff
@@ -0,0 +1,100 @@
+#include "restart_handler_unix.hpp"
+#include <fstream>
+#include <logger.hpp>
+
+#include <chrono>
+#include <thread>
+
+namespace restart_handler
+{
+
+    std::vector<char*> RestartHandler::cmd_line;
+
+    bool UsingSystemctl()
+    {
+        return (0 == std::system("which systemctl > /dev/null 2>&1") && nullptr != std::getenv("INVOCATION_ID"));
+    }
+
+    boost::asio::awaitable<module_command::CommandExecutionResult> RestartWithSystemd()
+    {
+        LogInfo("Systemctl restarting wazuh agent service.");
+        if (std::system("systemctl restart wazuh-agent") != 0)
+        {
+            co_return module_command::CommandExecutionResult {module_command::Status::IN_PROGRESS,
+                                                              "Systemctl restart execution"};
+        }
+        else
+        {
+            LogError("Failed using systemctl.");
+            co_return module_command::CommandExecutionResult {module_command::Status::FAILURE,
+                                                              "Systemctl restart failed"};
+        }
+    }
+
+    void StopAgent()
+    {
+        const int timeout = 30;
+        time_t start_time = time(nullptr); // Record the start time to track the timeout duration
+
+        pid_t pid = getppid();
+
+        // Shutdown Gracefully
+        kill(pid, SIGTERM);
+
+        while (true)
+        {
+            if (kill(pid, 0) != 0)
+            {
+                LogInfo("Agent gracefully stopped.");
+                break;
+            }
+
+            if (difftime(time(nullptr), start_time) > timeout)
+            {
+                LogError("Timeout reached! Forcing agent process termination.");
+                kill(pid, SIGKILL);
+            }
+
+            std::this_thread::sleep_for(std::chrono::seconds(1));
+        }
+    }
+
+    boost::asio::awaitable<module_command::CommandExecutionResult> RestartWithFork()
+    {
+        pid_t pid = fork();
+
+        if (pid < 0)
+        {
+            LogError("Fork failed.");
+        }
+        else if (pid == 0)
+        {
+            // Child process
+            StopAgent();
+
+            LogInfo("Starting wazuh agent in a new process.");
+
+            if (execve(RestartHandler::cmd_line[0], RestartHandler::cmd_line.data(), nullptr) == -1)
+            {
+                LogError("Failed to spawn new Wazuh agent process.");
+            }
+        }
+
+        co_return module_command::CommandExecutionResult {module_command::Status::IN_PROGRESS,
+                                                          "Pending restart execution"};
+    }
+
+    boost::asio::awaitable<module_command::CommandExecutionResult> RestartHandler::RestartCommand()
+    {
+
+        if (UsingSystemctl())
```

Comment:
Shouldn't you do something similar in macOS? Have you tested in macOS?

Reply:
We should probably do that, but we could address it in another issue. I haven't tested it on macOS.

Reply:
In that case, why don't you create a `restart_handler_lin.cpp` and `restart_handler_mac.cpp` (similar to win)?

Reply:
> Shouldn't you do something similar in macOS (launchctl command)? Have you tested in macOS?
I think it's very likely that it will work for Linux and macOS. 
As you mentioned, we should add a conditional for launchctl, so we don't need another file. 

> In that case, why don't you create a restart_handler_lin.cpp and restart_handler_mac.cpp (similar to win)?
I believe this could be addressed in the macOS issue if needed.

Reply:
Ok, so please upload a test of this working on macOS to complete evidence.

Reply:
Sure, I tested it [here](https://github.com/wazuh/wazuh-agent/actions/runs/12990813666/job/36226889350?pr=547#:~:text=35/39%20Test,1658).